src/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs

```diff
@@ -1915,17 +1915,21 @@ public bool IsCompleted
                 }
             }
 
-            internal unsafe DisconnectAsyncResult(HttpListener httpListener, ulong connectionId)
+            internal DisconnectAsyncResult(HttpListener httpListener, ulong connectionId)
             {
                 if (NetEventSource.IsEnabled) NetEventSource.Info(this, $"HttpListener: {httpListener}, ConnectionId: {connectionId}");
                 _ownershipState = 1;
                 _httpListener = httpListener;
                 _connectionId = connectionId;
 
-                // we can call the Unsafe API here, we won't ever call user code
-                _requestQueueBoundHandle = httpListener.RequestQueueBoundHandle;
-                _nativeOverlapped = _requestQueueBoundHandle.AllocateNativeOverlapped(s_IOCallback, state: this, pinData: null);
-                if (NetEventSource.IsEnabled) NetEventSource.Info($"DisconnectAsyncResult: ThreadPoolBoundHandle.AllocateNativeOverlapped({_requestQueueBoundHandle}) -> {_nativeOverlapped->GetHashCode()}");
+                // HttpListener.Stop() will reset its RequestQueueBoundHandle. Take a reference to the one we
+                // allocate on to ensure the OVERLAPPED structure is freed on the proper request queue if the
+                // disconnect I/O completion comes in after Stop() is called.
+                _disconnectRequestQueue = httpListener.RequestQueueBoundHandle;
```

Comment:
This is not queue, is it? Can we have more than one operation pending?

Reply:
Every connection might have two pending operations (one for disconnect, one for data I/O), and you can have multiple connections. When the operations complete, their completions are put into this queue which is then dequeued and surfaced through to the async result as a callback.

Reply:
Seems like it'd be better to just name it the same as where it came from: _requestQueueBoundHandle.

Reply:
How about `_overlappedAllocator`? We aren't keeping the local around to use as a queue, only to deallocate something.

Reply:
Why innovate in the name, rather than just naming it the same as what it is?  The current code uses "RequestQueueBoundHandle" for all accesses... why not just call the field "_requestQueueBoundHandle"?

Reply:
(I'm fine with David's suggestion of adding "Copy" to the end of the name as well.)

Reply:
"Copy" is equally confusing to me as I'd read it as a new instance rather than just a reference.

Reply:
Then I'd just stick with the original name: _requestQueueBoundHandle.

Reply:
I'm fine with that.